src/AudioCaptureImpl_WASAPI.cpp

```diff
@@ -1,25 +1,631 @@
+
+        auto devices{GetAudioDeviceList(enumerator)};
+        bool useLoopback{true};
+        std::string deviceName;
+
+        IMMDevice* device{nullptr};
+
+        if (_currentAudioDeviceIndex == -1 || _currentAudioDeviceIndex >= devices.size())
+        {
+            // Get the default render endpoint for opening it as a loopback device.
+            result = enumerator->GetDefaultAudioEndpoint(eRender, eConsole, &device);
+            if (FAILED(result))
+            {
+                poco_error_f1(_logger, "IMMDeviceEnumerator::GetDefaultAudioEndpoint failed: result = 0x%08?x", result);
+                break;
+            }
+
+            deviceName = _defaultDeviceName;
+        }
+        else
+        {
+            // Get a device by its ID according to the currently selected index.
+            useLoopback = devices.at(_currentAudioDeviceIndex).IsRenderDevice();
+            deviceName = devices.at(_currentAudioDeviceIndex).FriendlyName();
+            result = enumerator->GetDevice(devices.at(_currentAudioDeviceIndex).DeviceId(), &device);
+
+            if (FAILED(result))
+            {
+                poco_error_f1(_logger, "IMMDeviceEnumerator::GetDevice failed: result = 0x%08?x", result);
+                break;
+            }
+        }
+
+        LPWSTR deviceID{nullptr};
+        result = device->GetId(&deviceID);
+        if (FAILED(result))
+        {
+            poco_error_f1(_logger, "IMMDevice::GetId failed: result = 0x%08?x", result);
+        }
+        else
+        {
+            _currentCaptureDeviceId = UnicodeToString(deviceID);
+        }
+
+        if (!OpenAudioDevice(device, useLoopback))
+        {
+            _isCapturing = false;
+        }
+
+        poco_information_f3(_logger, "Audio device opened: %s (channels: %hu, loopback: %b)", deviceName, _channels, useLoopback);
+
+        while (_isCapturing && !_restartCapturing)
+        {
+            try
+            {
+                _fillBufferEvent.tryWait(500);
+            }
+            catch (Poco::TimeoutException& ex)
+            {
+                poco_debug(_logger, "FillBuffer event timeout, proceeding to flush buffer or abort.");
+            }
+
+            if (!_isCapturing)
+            {
+                break;
+            }
+
+            UINT32 packetLength;
+
+            _audioCaptureClient->GetNextPacketSize(&packetLength);
+            while (packetLength != 0)
+            {
+                BYTE* data;
+                UINT32 framesAvailable;
+                DWORD flags;
+
+                result = _audioCaptureClient->GetBuffer(&data, &framesAvailable, &flags, nullptr, nullptr);
+                if (FAILED(result))
+                {
+                    poco_error_f1(_logger, "IAudioCaptureClient::GetBuffer failed: result = 0x%08?x", result);
+                    _isCapturing = false;
+                    break;
+                }
+
+                if (flags & AUDCLNT_BUFFERFLAGS_SILENT)
```

Comment:
If there's silence, we could probably supply projectM with a buffer containing zeroes to flatten the waveform. Currently, if a device only has silence (not exactly sure when this happens, maybe if a mic is muted), projectM will keep displaying the last buffer contents as a frozen waveform. No harm being done, but a flat waveform would possibly look better.